### ISSUE TITLE
Validation run on string instead of object

### ISSUE THREAD

**Issue by samekmichal** [apigee-127/swagger-test-templates#94]
Currently, i am experimenting with project `swagger-node`, which utilizes this project - so thats why I report here.

I am having troubles validate payload responses against defined schemas. The source of the problem is in the test file, where the response body (json string) is not parsed into an object.

For illustration, excerpt from file in `/test/api/client/...`

``` js
...
function(error, res, body) {
  if (error) return done(error);

  res.statusCode.should.equal(200);

  // body = '{"message": "here i am"}'
  validator.validate(body, schema).should.be.true;
  done();
});
...
```

Why the response payload does not get converted to an object? When I manually adjust the code the validation succeeds...

``` js
validator.validate(JSON.parse(body), schema).should.be.true;
```

Shouldnt the template be adjust to reflect this?

**Comment by noahdietz** [comment]
@samekmichal you're right, the body should be parsed out to a JSON object like you've provided in your example.

What needs to happen here is the templates should be adjusted to include the change and all of the diff files for testing must be updated to reflect the change...would you submit a PR for this? it's a simple change, but lots of files to change

**Comment by samekmichal** [comment]
OK, I will have a try.

**Comment by noahdietz** [comment]
@Remco75 have you noticed this at all? It seems weird to me that this looks to be a bug but no one else has reported having trouble with it.

I dug around the `request` [docs](https://www.npmjs.com/package/request#custom-http-headers) a bit and confirmed that when the data must be accessed they do indeed `JSON.parse()` the `body` param of the callback.

**Comment by Remco75** [comment]
@noahdietz I will look into it tomorrow. I did not notice it yet... strange indeed. Let's investigate shortly, and decide

**Comment by noahdietz** [comment]
Also, this only pertains to use of `request` as the http agent. Using `supertest` references the `res.body` value for validation, as it doesn't have a third `body` param.

**Comment by noahdietz** [comment]
Could this be a backend implementation detail? I'm looking at the source code for `request` and it looks like if `json` is being transferred it will parse the JSON object, store in the response object and even return it as the extra parameter ([`request` source reference](https://github.com/request/request/blob/master/request.js#L1037-L1048)). I'm kind of shooting from the hip here, just getting thoughts down.

**Comment by Remco75** [comment]
Dug into this a bit, and found 2 things:
According to the [docs](https://www.npmjs.com/package/request#requestoptions-callback) we should set json to true and then the body will be parsed, and the response is JSON, and it will set the correct header.

We do it slightly different, we set the body _inside_ the json and set the application/json header ourselves.

To at least better catch these errors, I can make a PR for that. I you still mangage to get a JSON string after this you are doing something weird ;-)

So, to sum it up: I don't think we should fix this, it's not a bug, but wel should improve the code: I will make a PR if we want this

(so let me know)

**Comment by noahdietz** [comment]
@Remco75 so what kind of change are you proposing? Just adding `json: true` to request options?

Even if what we do currently "works", if there is a way to make it more reliable/idiomatic, I'm ok with that, so go ahead with a PR if you've got time.

**Comment by Remco75** [comment]
Ok, will try to find time to make a PR tomorrow or Wednesday. But again, this does not seem like a true bug to me.

**Comment by noahdietz** [comment]
@Remco75 no pressure, if this doesn't make the cut for a new release, I'm not too worried.

**Comment by Remco75** [comment]
hmm, this is getting some serious work for a bug that might not even be a bug. 
Im skipping this one, at least for the release.

**Comment by baynezy** [comment]
I am happy to do this work as this is a big problem for me. I would just like some guidance on what you were both imagining. `json: true` definitely fixes my issues, but I want to understand if that is what you had in mind?

**Comment by baynezy** [comment]
As you can see if you run this it definitely does not return an object by default:-

    var request = require("request");

    request({
            url: 'https://api.github.com/users/baynezy',
            method: 'GET',
            headers: {
                'User-Agent': 'baynezy'
            }
        },
        function(error, res, body) {
            if (error) return done(error);

            console.log(typeof body);
            console.log(body);
        });

**Comment by noahdietz** [comment]
In looking at the request docs more and the templates, I'm wondering how what we do currently even works.

If request data is given for the `body`, it populates the object in the `request.options.json` property...this should just be a boolean flag that toggles the parsing of JSON by request. The data should really be put in the `request.options.body`.

@baynezy if you want to make the changes, I'll accept a PR.
* set `json: true`, this should only happen when the produces/consumes is `application/json`
* put the body parameter objects in the `body` property. This should be changing just the property name in the templates and diff files.

**Comment by baynezy** [comment]
@noahdietz - I am not sure you can limit it to just `application\json`. [Vendor specific media types](https://en.wikipedia.org/wiki/Media_type#Registration_trees) are quite prevalent e.g. `application/vnd.github.v3+json`.

I can go with just implementing `application\json` for now, but there will be more work needed.

**Comment by noahdietz** [comment]
@baynezy good point
would it be too hacky to right a handlebars helper to check the consume/produces strings for `json`? If that sounds like a reasonable approach, go for it, then we can support `json` in general.
